integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonPreAggregateQueryRules.scala

```diff
@@ -0,0 +1,756 @@
+             &&
+             logicalRelation.relation.asInstanceOf[CarbonDatasourceHadoopRelation].carbonTable
+               .hasPreAggDataMap =>
+          val (carbonTable, tableName) = getCarbonTableAndTableName(logicalRelation)
+          // when table has child data map(pre aggregate table) then only plan will be transformed
+          if (!carbonTable.hasPreAggDataMap) {
+            isValidPlan = false
+          }
+          if (isValidPlan) {
+            // if it is valid plan then extract the query columns
+            isValidPlan = extractQueryColumnsFromAggExpression(groupingExp,
+              aggregateExp,
+              carbonTable,
+              tableName,
+              list)
+          }
+          carbonTable
+        case _ =>
+          isValidPlan = false
+          null
+      }
+      // if plan is valid then update the plan with child attributes
+      if (isValidPlan) {
+        // getting all the projection columns
+        val listProjectionColumn = list
+          .filter(queryColumn => queryColumn.getAggFunction.isEmpty && !queryColumn.isFilterColumn)
+        // getting all the filter columns
+        val listFilterColumn = list
+          .filter(queryColumn => queryColumn.getAggFunction.isEmpty && queryColumn.isFilterColumn)
+        // getting all the aggregation columns
+        val listAggregationColumn = list.filter(queryColumn => !queryColumn.getAggFunction.isEmpty)
+        // create a query plan object which will be used to select the list of pre aggregate tables
+        // matches with this plan
+        val queryPlan = new QueryPlan(listProjectionColumn.asJava,
+          listAggregationColumn.asJava,
+          listFilterColumn.asJava)
+        // create aggregate table selector object
+        val aggregateTableSelector = new AggregateTableSelector(queryPlan, carbonTable)
+        // select the list of valid child tables
+        val selectedDataMapSchemas = aggregateTableSelector.selectPreAggDataMapSchema()
+        // if it doesnot match with any pre aggregate table return the same plan
+        if (!selectedDataMapSchemas.isEmpty) {
+          // sort the selected child schema based on size to select smallest pre aggregate table
+          val (aggDataMapSchema, carbonRelation) =
+            selectedDataMapSchemas.asScala.map { selectedDataMapSchema =>
+              val catalog = sparkSession.sessionState.catalog
+              val carbonRelation = catalog
+                .lookupRelation(TableIdentifier(selectedDataMapSchema.getRelationIdentifier
+                  .getTableName,
+                  Some(selectedDataMapSchema.getRelationIdentifier
+                    .getDatabaseName))).asInstanceOf[SubqueryAlias].child
+                .asInstanceOf[LogicalRelation]
+              (selectedDataMapSchema, carbonRelation)
+            }.sortBy(f => f._2.relation.asInstanceOf[CarbonDatasourceHadoopRelation].sizeInBytes)
+              .head
```

Comment:
Instead of using `}.sortBy(f => f._2.relation.asInstanceOf[CarbonDatasourceHadoopRelation].sizeInBytes).head`
use `.minBy(f => f._2.relation.asInstanceOf[CarbonDatasourceHadoopRelation].sizeInBytes)`